src/sonic-ctrmgrd/ctrmgr/kube_commands.py

```diff
@@ -0,0 +1,388 @@
+    if ret != 0:
+        log_debug("Labels read failed.")
+        return ret
+
+    print(json.dumps(node_labels, indent=4))
+    return 0
+    
+
+def is_connected(server=""):
+    """ Check if we are currently connected """
+    KUBELET_YAML = "/var/lib/kubelet/config.yaml"
+
+    if (os.path.exists(KUBELET_YAML) and os.path.exists(KUBE_ADMIN_CONF)):
+        with open(KUBE_ADMIN_CONF, 'r') as s:
+            d = yaml.load(s, yaml.SafeLoader)
+            d = d['clusters'] if 'clusters' in d else []
+            d = d[0] if len(d) > 0 else {}
+            d = d['cluster'] if 'cluster' in d else {}
+            d = d['server'] if 'server' in d else ""
+            if d:
+                o = urlparse(d)
+                if o.hostname:
+                    return not server or server == o.hostname
+    return False
+
+
+def func_is_connected(args):
+    """ Get connected state """
+    connected = is_connected()
+    print("Currently {} to Kube master".format(
+        "connected" if connected else "not connected"))
+    return 0
+
+
+def _take_lock():
+    """ Take a lock to block concurrent calls """
+    LOCK_FILE = "/var/lock/kube_join.lock"
+    lock_fd = None
+    try:
+        lock_fd = open(LOCK_FILE, "w")
+        fcntl.lockf(lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        log_debug("Lock taken {}".format(LOCK_FILE))
+    except IOError as e:
+        lock_fd = None
+        log_error("Lock {} failed: {}".format(LOCK_FILE, str(e)))
+    return lock_fd
+
+
+def _download_file(server, port, insecure):
+    """ Download file from Kube master to assist join as node. """
+    SERVER_ADMIN_URL = "https://{}/admin.conf"
+
+    if insecure:
+        r = urllib.request.urlopen(SERVER_ADMIN_URL.format(server),
+                context=ssl._create_unverified_context())
+    else:
+        r = urllib.request.urlopen(SERVER_ADMIN_URL.format(server))
+
+    (h, fname) = tempfile.mkstemp(suffix="_kube_join")
+    data = r.read()
+    os.write(h, data)
+    os.close(h)
+
+    # Ensure the admin.conf has given VIP as server-IP.
+    update_file = "{}.upd".format(fname)
+    cmd = 'sed "s/server:.*:{}/server: https:\/\/{}:{}/" {} > {}'.format(
+            str(port), server, str(port), fname, update_file)
+    (ret, _, err) = _run_command(cmd)
+
+    if ret != 0:
+        log_err("sed update of downloaded file failed with ret={}\n{}".
+                format(ret, err))
+        return ret
+
+    shutil.copyfile(update_file, KUBE_ADMIN_CONF)
+
+    _run_command("rm -f {} {}".format(fname, update_file))
+    log_debug("{} downloaded".format(KUBE_ADMIN_CONF))
+
+
+def _troubleshoot_tips():
+    """ log troubleshoot tips which could be handy,
+        when in trouble with join
+    """
+    msg = """
+if join fails, check the following
+
+a.  Ensure both master & node run same or compatible k8s versions
+
+b.  Check if this node already exists in master
+    Use 'sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf get nodes' to list nodes at master.
```

Comment:
the --kubeconfig `KUBE_ADMIN_CONF` download path is set to `/etc/sonic/kube_admin.conf `